src/common/webdav.ts

```diff
@@ -0,0 +1,110 @@
+// Copyright (C) The Arvados Authors. All rights reserved.
+//
+// SPDX-License-Identifier: AGPL-3.0
+
+export class WebDAV {
+    static create(config?: Partial<WebDAVDefaults>, createRequest?: () => XMLHttpRequest) {
+        return new WebDAV(config, createRequest);
+    }
+
+    defaults: WebDAVDefaults = {
+        baseUrl: '',
+        headers: {},
+    };
+
+    propfind = (url: string, config: PropfindConfig = {}) =>
+        this.request({
+            ...config, url,
+            method: 'PROPFIND'
+        })
+
+    put = (url: string, config: PutConfig = {}) =>
+        this.request({
+            ...config, url,
+            method: 'PUT'
+        })
+
+    copy = (url: string, { destination, ...config }: CopyConfig) =>
+        this.request({
+            ...config, url,
+            method: 'COPY',
+            headers: { ...config.headers, Destination: this.defaults.baseUrl + destination }
+        })
+
+    move = (url: string, { destination, ...config }: MoveConfig) =>
+        this.request({
+            ...config, url,
+            method: 'MOVE',
+            headers: { ...config.headers, Destination: this.defaults.baseUrl + destination }
+        })
+
+    delete = (url: string, config: DeleteConfig = {}) =>
+        this.request({
+            ...config, url,
+            method: 'DELETE'
+        })
+
+    private constructor(config?: Partial<WebDAVDefaults>, private createRequest = () => new XMLHttpRequest()) {
+        if (config) {
+            this.defaults = { ...this.defaults, ...config };
+        }
+    }
+
+    private request = (config: RequestConfig) => {
+        return new Promise<XMLHttpRequest>((resolve, reject) => {
```

Comment:
Why not using axios here? In axios config you can pass progress callback. See upload component for example